### PR TITLE
[ao] fixing quantized prelu workflow

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103455

Summary: https://github.com/pytorch/pytorch/issues/100654 noticed prelu
was not running its observers when the quantization flow was being run,
this was a bug which is now fixed and the relevant prelu tests also now
check for this. Also added a corrected observer for PReLU to
qconfig_mapping

Test Plan: python test/test_quantization.py TestStaticQuantizedModule.test_prelu

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: test/quantization/core/test_quantized_module.py [hunk 1]
```diff
@@ -1314,29 +1314,27 @@ def test_embedding_bag_api(self, num_embeddings, embedding_dim, num_offsets, set
...
+
+            # check that the weight makes sense
+            self.assertEqual(q_prelu.weight.dequantize(), f_prelu.weight, atol=.1, rtol=.1)
+            f_prelu.weight = torch.nn.Parameter(q_prelu.weight.dequantize())
+            qy = q_prelu(qx)
+            qy_ref = torch.quantize_per_tensor(
+                f_prelu(qx.dequantize()), q_prelu.scale, q_prelu.zero_point, dtype=torch.quint8
+            )
+            # check that the output makes sense
+            self.assertEqual(qy, qy_ref, atol=.1, rtol=.1)
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_module.py:1337]: should this be closer?

File: test/quantization/core/test_quantized_module.py [hunk 2]
```diff
@@ -1314,29 +1314,27 @@ def test_embedding_bag_api(self, num_embeddings, embedding_dim, num_offsets, set
...
+
+            f_prelu = torch.nn.PReLU(num_parameters=num_parameters)
+            f_prelu.weight = torch.nn.Parameter(torch.randn(num_parameters).abs())
+            f_prelu.qconfig = torch.ao.quantization.get_default_qconfig('qnnpack')
+            f_prelu.activation_post_process = f_prelu.qconfig.activation()
+            f_prelu.activation_post_process(f_prelu(x))
+            q_prelu = nnq.PReLU.from_float(f_prelu)
+
+            # check that the weight makes sense
+            self.assertEqual(q_prelu.weight.dequantize(), f_prelu.weight, atol=.1, rtol=.1)
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_module.py:1330]: can we just compare the q_relu.weight.dequantize() also with a quantized f_prelu.weight and check that they are the same?

File: test/quantization/core/test_quantized_module.py (hunk 1 above)
Comment (HDCharles) [reply test/quantization/core/test_quantized_module.py:1337]: subject to numerics weirdness, not sure how the different backends are doing the requantization,

File: test/quantization/core/test_quantized_module.py (hunk 2 above)
Comment (HDCharles) [reply test/quantization/core/test_quantized_module.py:1330]: we are doing that, but they aren't the same unless the original weights fall exactly on the quantization points.

Comment (HDCharles) [reply test/quantization/core/test_quantized_module.py:1330]: oh you want me to quantize it first? I can do that.

File: torch/ao/nn/quantized/modules/activation.py [hunk 3]
```diff
@@ -271,6 +271,11 @@ def from_float(cls, mod):
         qprelu = cls(float(scale), int(zero_point), mod.num_parameters)
         float_wt = mod.weight.float()
         observer = mod.qconfig.weight()
+        observer(float_wt.unsqueeze(-1))
```
Comment (jerryzh168) [inline torch/ao/nn/quantized/modules/activation.py:274]: why do we need to unsqueeze?

Comment (HDCharles) [reply torch/ao/nn/quantized/modules/activation.py]: because if you try to use a per_channel weight observer on it, then it will throw an error because y = torch.flatten(y, start_dim=1) doesn't work when dim=1 is out of bounds. Its why (https://github.com/pytorch/pytorch/pull/103455/commits/9c6927f1ef9b847b26ecee9575227d27491c7662) didn't work.

Comment (jerryzh168) [reply torch/ao/nn/quantized/modules/activation.py:274]: does this mean this layer does not support per channel weight quantization? I think in that case we should fix the place where people requested the per channel weight quantization, instead of here

File: torch/ao/quantization/qconfig_mapping.py [hunk 4]
```diff
@@ -92,6 +93,7 @@ def _get_default_qconfig_mapping(is_qat: bool, backend: str, version: int) -> QC
         .set_object_type(torch.nn.functional.conv_transpose3d, qconfig_transpose) \
         .set_object_type(torch.nn.functional.layer_norm, qconfig_layernorm) \
         .set_object_type(torch.nn.LayerNorm, qconfig_layernorm) \
+        .set_object_type(torch.nn.PReLU, default_quint8_weight_qconfig) \
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig_mapping.py:96]: do you still need to unsqueeze weight with this configuration?

**Review by jerryzh168 — APPROVED** [review]

File: torch/ao/quantization/qconfig_mapping.py (hunk 4 above)
Comment (HDCharles) [reply torch/ao/quantization/qconfig_mapping.py:96]: no, you don't.
